### PR TITLE
[CARBONDATA-188] Compress CSV file before loading

### PR DESCRIPTION
**Pull request by jackylk**
Currently when loading CarbonData file using Spark Dataframe API, it will firstly save as CSV file then load to CarbonData file. 

Sometimes CSV requires a lot of disk space,  in this PR, instead of saving as CSV text file, it will save a compressed CSV file, then load to CarbonData. 

In my laptop, when loading 1 million records, the disk space required for CSV file is reduced 4~5 times.

And in this PR, the csv extension name check is removed for loading, user can load CSV file with any file extension.

### CODE REVIEW

File: processing/src/main/java/org/apache/carbondata/processing/csvreaderstep/UnivocityCsvParser.java [hunk 1]
```diff
@@ -112,25 +116,29 @@ private void initializeReader() throws IOException {
...
+
+      DataInputStream dataInputStream = FileFactory.getDataInputStream(path, fileType, bufferSize);
+
+      // if start offset is not 0 then reading then reading and ignoring the extra line
+      if (startOffset != 0) {
+        LineReader lineReader = new LineReader(dataInputStream, 1);
+        startOffset += lineReader.readLine(new Text(), 0);
+      }
+      inputStreamReader = new BufferedReader(new InputStreamReader(
+          new BoundedDataStream(dataInputStream, endOffset - startOffset)));
```
Comment (Zhangshunyu) [inline processing/src/main/java/org/apache/carbondata/processing/csvreaderstep/UnivocityCsvParser.java]: Can not find class BoundedDataStream

Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/csvreaderstep/UnivocityCsvParser.java]: Forgot to add.
Added now

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 2]
```diff
@@ -657,6 +657,8 @@ object CarbonDataRDDFactory extends Logging {
           val filePaths = carbonLoadModel.getFactFilePath
           hadoopConfiguration.set("mapreduce.input.fileinputformat.inputdir", filePaths)
           hadoopConfiguration.set("mapreduce.input.fileinputformat.input.dir.recursive", "true")
+          hadoopConfiguration.set("io.compression.codecs",
+            "org.apache.hadoop.io.compress.GzipCodec")
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: This configuration is only for compression file.
Please check whether if it is a compression file.

File: integration/spark/src/main/scala/org/apache/carbondata/spark/util/GlobalDictionaryUtil.scala [hunk 3]
```diff
@@ -364,6 +364,7 @@ object GlobalDictionaryUtil extends Logging {
       .option("escape", carbonLoadModel.getEscapeChar)
       .option("ignoreLeadingWhiteSpace", "false")
       .option("ignoreTrailingWhiteSpace", "false")
+      .option("codec", "gzip")
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/util/GlobalDictionaryUtil.scala]: Please check whether if it is a compression file.

File: integration/spark/src/main/scala/org/apache/carbondata/spark/csv/CarbonTextFile.scala [hunk 4]
```diff
@@ -36,6 +36,8 @@ private[csv] object CarbonTextFile {
     val hadoopConfiguration = new Configuration(sc.hadoopConfiguration)
     hadoopConfiguration.setStrings(FileInputFormat.INPUT_DIR, location)
     hadoopConfiguration.setBoolean(FileInputFormat.INPUT_DIR_RECURSIVE, true)
+    hadoopConfiguration.set("io.compression.codecs", "org.apache.hadoop.io.compress.GzipCodec")
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/csv/CarbonTextFile.scala]: Please check whether if it is a compression file.

**Comment by QiangCai** [comment]
Please modify the readCSVFile method in GraphExecutionUtil.java to support  gzip file.

**Comment by jackylk** [comment]
@QiangCai please check

**Comment by QiangCai** [comment]
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/68/

File: processing/src/main/java/org/apache/carbondata/processing/csvreaderstep/UnivocityCsvParser.java [hunk 5]
```diff
@@ -112,25 +116,28 @@ private void initializeReader() throws IOException {
...
-    // if start offset is not 0 then reading then reading and ignoring the extra line
-    if (startOffset != 0) {
-      LineReader lineReader = new LineReader(dataInputStream, 1);
-      startOffset += lineReader.readLine(new Text(), 0);
+
+    String path = this.csvParserVo.getBlockDetailsList().get(blockCounter).getFilePath();
+    FileType fileType = FileFactory.getFileType(path);
+
+    DataInputStream dataInputStream =
+        FileFactory.getDataInputStream(path, fileType, bufferSize);
```
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/csvreaderstep/UnivocityCsvParser.java]: For csv file ,  DataInputStream need startOffSet

**Comment by QiangCai** [comment]
please modify test case MultiFilesDataLoadingTestCase.
I think the result should be 10 rows now.

**Comment by QiangCai** [comment]
CI PASS
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/103/

LGTM

**Pull request closed without merging**